cpp/src/arrow/dataset/partition_test.cc

```diff
@@ -406,6 +445,18 @@ TEST_F(TestPartitioning, HivePartitioningFormat) {
            equal(field_ref("beta"), literal("hello"))));
 }
 
+TEST_F(TestPartitioning, FilenamePartitioningFormat) {
+  partitioning_ = std::make_shared<FilenamePartitioning>(
+      schema({field("alpha", int32()), field("beta", utf8())}));
+
+  written_schema_ = partitioning_->schema();
+
+  AssertFormat(and_(equal(field_ref("alpha"), literal(0)),
+                    equal(field_ref("beta"), literal("hello"))),
+               "", "0_hello_");
```

Comment:
What happens if instead of `literal("hello")` I pass `literal("foo_bar")` or `literal("foo/bar")`?

Reply:
With this code, if we have `literal(foo/bar)`,  then the prefix string is expected to be `"0_foo/bar_"`,  with `literal(foo_bar)`, the prefix string is expected to be `"0_foo_bar_"`.

Reply:
This case is actually not handled anywhere as of now. As suggested by @westonpace, we can URI encode the segments here to handle this case, I will create a JIRA issue for a follow-up PR on this.